src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs

```diff
@@ -881,4 +883,152 @@ protected static async Task<ImmutableArray<string>> GetAllMatchingGlobalAliasNam
+            {
+                var (symbol, symbolSet) = args;
+                var name = symbol.Name;
+
+                var shouldProcess = await ShouldProcessDocumentAsync(document, name, cancellationToken).ConfigureAwait(false);
+                if (shouldProcess)
+                {
+                    callback(document);
+                }
+            },
+            consumeItems: static async (documents, args, cancellationToken) =>
+            {
+                var (symbol, symbolSet) = args;
+                await foreach (var document in documents)
+                {
+                    await FindImplicitSymbolsInDocumentAsync(symbol, symbolSet, document, cancellationToken).ConfigureAwait(false);
+                }
+            },
+            args: (symbol, symbolSet),
+            cancellationToken).ConfigureAwait(false);
+
+        symbolBuilder.AddRange(symbolSet);
+
+        async IAsyncEnumerable<Document> GetCandidateDocumentsAsync()
+        {
+            var name = symbol.Name;
+
+            foreach (var project in solution.Projects)
+            {
+                var allDocuments = project.GetAllRegularAndSourceGeneratedDocumentsAsync(cancellationToken);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                await foreach (var document in allDocuments)
+                {
+                    yield return document;
+                }
+            }
+        }
+
+        static async ValueTask<bool> ShouldProcessDocumentAsync(Document document, string name, CancellationToken cancellationToken)
+        {
+            if (!document.SupportsSemanticModel)
+                return false;
+
+            var treeIndex = await document.GetSyntaxTreeIndexAsync(cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (treeIndex is null)
+                return false;
+
+            return treeIndex.ProbablyContainsIdentifier(name);
+        }
+    }
+
+    private static async ValueTask FindImplicitSymbolsInDocumentAsync(TSymbol symbol, PooledHashSet<ISymbol> symbolSet, Document document, CancellationToken cancellationToken)
+    {
+        var semanticModel = await document!.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+        if (cancellationToken.IsCancellationRequested)
+            return;
+
+        Debug.Assert(semanticModel is not null);
+
+        var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        if (root is null)
+            return;
+
+        var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+        // Instead of getting all the declared symbols from the document, invoke this semantic operation in every node we encounter
+        foreach (var descendantNode in root.DescendantNodes())
+        {
+            var documentSymbol = semanticModel.GetDeclaredSymbol(descendantNode, cancellationToken);
```

Comment:
This is a non starter. You are trying to bind every node in the document. Again, this breaks the design of find references.